src/main/scala/com/gu/subscriptions/cas/directives/ProxyDirective.scala

```diff
@@ -134,38 +138,80 @@ trait ProxyDirective extends Directives with ErrorRoute with LazyLogging {
         // return error: "Mandatory data missing from request"
   }
 
-  def zuoraRoute(subsReq: SubscriptionRequest): Route = zuoraDirective(subsReq) { (activation, subscriptionName) =>
-    val validSubscription = subscriptionService.getValidSubscription(Name(subscriptionName.get.trim.dropWhile(_ == '0')), subsReq.password.getOrElse(""))
 
-    validSubscription.onFailure {
-      case t: Throwable =>
-        logger.error(s"Failed getting Zuora subscription ${t.getMessage} ${subsReq.subscriberId}")
-        throw t
-    }
+  def processEmergencyToken(req: SubscriptionRequest): Option[SubscriptionExpiration] = {
+    import com.gu.subscriptions.cas.model.TokenPayloadOps._
+
+    req.subscriberId.flatMap { rawSubscriberId1 =>
+      val subsId = rawSubscriberId1.trim.toUpperCase
+
+      def processPayload(payloadResult: PayloadResult) = payloadResult match {
+        case Valid(payload) =>
+          logger.info(s"subscriber id:'$subsId' was created on ${payload.creationDate}")
+
+          val subExp = SubscriptionExpiration(
+            expiryDate = payload.expiryDate,
+            expiryType = ExpiryType.SUB,
+            subscriptionCode = Some(payload.subscriptionCode),
+            provider = Some(Configuration.EmergencyTokens.prefix)
+          )
+          Some(subExp)
+        case _ => None
+      }
 
-    onSuccess(validSubscription) {
-      case Some(subscription: Subscription[Paid]) =>
-        if (activation) { subscriptionService.updateActivationDate(subscription) }
-        // TODO ASAP - if deviceId and appId are provided:
-          // upsert a record in DynamoDB
-        //Since the dates are in PST, we want to make sure that we don't cut any subscription one day short
-        complete(SubscriptionExpiration(subscription.termEndDate.plusDays(1).toDateTimeAtStartOfDay(), ExpiryType.SUB))
-      case _ if subscriptionName.get.startsWith("A-S") =>
-        //no point going to CAS if this is a Zuora sub
-        notFound
-      case _ =>
-        //Go to CAS legacy /subs route
-        reject
+      if (!subsId.startsWith(emergencyTokens.prefix)) {
+        None
+      } else {
+        logger.info(s"EMERGENCY PROVIDER triggered for subscriber id:'$subsId' password: '${req.password} deviceId:${req.deviceId}")
+
+        Try(emergencyTokens.decoder.decode(subsId)) match {
+
+          case Success(payloadResult) =>
+            logger.info(s"subscriber id:'$subsId' resolves to $payloadResult")
+            processPayload(payloadResult)
+
+          case Failure(cause) =>
+            logger.error(s"error decoding token $subsId :  $cause")
+            None
+        }
+      }
     }
   }
 
+  def processSub(subsReq: SubscriptionRequest): Route = {
+    processEmergencyToken(subsReq).map(complete(_)) getOrElse {
```

Comment:
took me a minute to realise complete is spray's equivalent of resolving a future http://spray.io/documentation/1.2.4/spray-routing/route-directives/complete/